---
title: "Analysis of results randomset2"
output: html_document
date: "2024-10-09"
---

```{r setup, include = FALSE}
# Set working directory
knitr::opts_knit$set(root.dir = '~/Documents/Projects/resistance_is_futile/amr-2024-team-prasad-feldgarden')
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
```

# Introduction

This notebook compares the results obtained using the `scanner.R` and `DGW.py` scripts. The objective is to assess whether both tools are yielding the same results, and if that is not the case, identify areas of opportunity to improve the scripts. In addition, the outputs from the scripts will be compared to the mutations found for those genomes in MicroBIGG-E.

# Setup

A random set of 30,000 genome accession IDs for *Acinetobacter baumannii*, *Pseudomonas aeruginosa* and *Klebsiella pneumoniae* (10,000 isolates for each) were obtained using NCBI's `datasets` as described [here](https://github.com/NCBI-Codeathons/amr-2024-team-prasad-feldgarden/tree/main/data/randomset1#test-on-a-random-set-of-assemblies).


## MicroBIGG-E subset

A subset of the data available in MicroBIGG-E for the selected isolates was obtained from BigQuery and stored in the repository [here](https://github.com/NCBI-Codeathons/amr-2024-team-prasad-feldgarden/blob/main/results/microbigge_subset_randomset2.csv). 

## DGW results

DGW was ran as described [here](https://github.com/NCBI-Codeathons/amr-2024-team-prasad-feldgarden/tree/main/script), and the output for each taxon is shown in individual tables in the repository [here](https://github.com/NCBI-Codeathons/amr-2024-team-prasad-feldgarden/tree/main/results). These results are going to be compared against the MicroBIGG-E subset in the following section. 

## DGW results: `scanner.R` and `DGW.py`

The scripts were ran as described [here](https://github.com/NCBI-Codeathons/amr-2024-team-prasad-feldgarden/tree/main/data/resis_and_sens_Kp_and_Pa).

## Combining output files

Individual output files for each genome screened were combined using the `results_combined.R` script ([here](https://github.com/NCBI-Codeathons/amr-2024-team-prasad-feldgarden/blob/3e9bb4e9dc2196f7f8f5ad2c5055133ebd35a87d/results/scanner.R_AST/results_combined.R)).

The combined output files from each script (`DGW.py` and `scanner.R`) were combined and modeled running the `dgw_combined_outputs.sql` model ([here](https://github.com/NCBI-Codeathons/amr-2024-team-prasad-feldgarden/blob/main/results/script_comparison/dgw_combined_outputs.sql)) in `DuckDB`. 

# Analysis of results

```{r}
# Install packges
library(tidyverse)
library(dplyr)


# Import data

## Microbigg-E subset
microbigge <-  read_csv("results/microbigge_subset_randomset2.csv")

## Combined and modeled output from both scanner.R and DGW.py scripts
dgw_combined_outputs <- read_csv("results/script_comparison/dgw_combined_outputs_randomset2.csv")

```

Let's first prepare the data to establish the comparisons/analysis.


### How many lesions are detected by each script?
  
  - `DWG.py` identified 8377 lesions
  - `scanner.R` identified 7405 lesions
  
```{r}
dgw_combined_outputs %>% 
  summarize(mutations_p = n_distinct(p_mutation_id, na.rm = TRUE),
            mutations_r = n_distinct(r_mutation_id, na.rm = TRUE)
  )
```
  
### How many lesions are found by both scripts? 

  - `DWG.py` and `scanner.R` identified the lesions in the same 7593 genomes
    - 5278 (69.5%) of those report the same lesions
  
```{r}
dgw_combined_outputs %>% 
  filter(!is.na(p_mutation_id) & !is.na(r_mutation_id)) %>% 
  summarize(in_both = n_distinct(record_id, na.rm = TRUE),
            same = n_distinct(case_when(p_mutation_id == r_mutation_id ~ record_id), na.rm = TRUE)
  )
```

### Which records have mutations identified by both scripts, but with different lesions identified by each script? 

- There were 2315 (30.5%) records where a mutation was identified by both `DGW.py` and `scanner.R`, but the lesions identified by each script were different
  - Of these, 220 (9.5%) records also reported mutations in Microbigg-E, which were not the same as the ones reported by the DGW scripts
  
```{r}
DT::datatable(dgw_combined_outputs %>% 
  filter(!is.na(p_mutation_id)  & !is.na(r_mutation_id) & p_mutation_id != r_mutation_id ) %>% 
  select(asm_acc, p_element_symbol, r_element_symbol, mb_element_symbol, p_lesion, r_lesion, mb_lesion)
)
```
    
    
### How many lesions are detected only by `DGW.py`?

- `DGW.py` identified 1035 records that `scanner.R` did not identify.
  - 241 (23.3%) of those records was also present in Microbigg-E

```{r}
DT::datatable(dgw_combined_outputs %>% 
  filter(!is.na(p_mutation_id) & is.na(r_mutation_id)) %>% 
  select(asm_acc, p_element_symbol, mb_element_symbol, p_lesion, mb_lesion)
  )

```

### How many lesions are detected only by `scaner.R`? 

- `scanner.R` identified 60 records that `DGW.py` did not identify
  - 4 (6.7%) of those mutations were present in Microbigg-E

```{r}
DT::datatable(dgw_combined_outputs %>% 
  filter(!is.na(r_mutation_id) & is.na(p_mutation_id)) %>% 
  select(asm_acc, r_element_symbol, mb_element_symbol, r_lesion, mb_lesion)
  )

```

## How many of the lesions detected by each script are in Microbigg-E?

- Of the 6406 lesions found in Microbigg-E
  - `DGW.py` identified 1606 (25.1%)
  - `scanner.R` identified 1369 (21.4%) (of which 1365 were also identified by `DGW.py`)

```{r}
DT::datatable(dgw_combined_outputs %>% 
  summarize(in_microbigge = n_distinct(case_when(!is.na(mb_element_symbol) ~ record_id), na.rm = TRUE),
            mb_p = n_distinct(case_when(!is.na(mb_element_symbol) & !is.na(p_element_symbol) ~ record_id), na.rm = TRUE),
            mb_r = n_distinct(case_when(!is.na(mb_element_symbol) & !is.na(r_element_symbol) ~ record_id), na.rm = TRUE),
            mb_combined = n_distinct(case_when(!is.na(mb_element_symbol) & (!is.na(p_element_symbol) | !is.na(r_element_symbol)) ~ record_id), na.rm = TRUE),
            mb_in_both = n_distinct(case_when(!is.na(mb_element_symbol) & !is.na(p_element_symbol) & !is.na(r_element_symbol) ~ record_id), na.rm = TRUE)
  )
)
```

## How many of the lesions detected by each script are NOT in Microbigg-E? (i.e. new mutations)

- The DGW scripts identified 7078 lesions that were not present in Microbigg-E
  - `DGW.py` identified 7022 (99.2%) new lesions
  - `scanner.R` identified 6284 (88.8%) new lesions
  - 6228 (88%) of the new lesions were identified by both scripts
  
```{r}
DT::datatable(dgw_combined_outputs %>% 
  summarize(new_p = n_distinct(case_when(is.na(mb_element_symbol) & !is.na(p_element_symbol) ~ record_id), na.rm = TRUE),
            new_r = n_distinct(case_when(is.na(mb_element_symbol) & !is.na(r_element_symbol) ~ record_id), na.rm = TRUE),
            new_combined = n_distinct(case_when(is.na(mb_element_symbol) & (!is.na(p_element_symbol) | !is.na(r_element_symbol)) ~ record_id), na.rm = TRUE),
            new_in_both = n_distinct(case_when(is.na(mb_element_symbol) & !is.na(p_element_symbol) & !is.na(r_element_symbol) ~ record_id), na.rm = TRUE)
  )
)
```

```{r, echo = FALSE}
union(dgw_combined_outputs %>% 
  mutate(result = case_when(is.na(mb_element_symbol) ~ "New",
                            !is.na(mb_element_symbol) ~ "In Microbigg-E"
                            )) %>% 
  group_by(result) %>% 
  summarize(script = "DWG.py",
            mutations = n_distinct(record_id))
  ,
dgw_combined_outputs %>% 
  mutate(result = case_when(is.na(mb_element_symbol) ~ "New",
                            !is.na(mb_element_symbol) ~ "In Microbigg-E"
                            )) %>% 
  group_by(result) %>% 
  summarize(script = "scanner.R",
            mutations = n_distinct(record_id))
) %>% 
  ggplot(., aes(fill=result, y=mutations, x=script)) + 
    geom_bar(position="stack", stat="identity", width=0.7) +
    scale_fill_manual(name="",values=c('#68a1b3', '#a65005')) +
    theme(
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.border = element_blank(),
      panel.background = element_blank(),
      text = element_text(size = 16)) +
    ylab("# of Mutations") + xlab("Script") +
    scale_x_discrete(labels = function(x) stringr::str_wrap(x, width = 20))
  
```


### Which genes have different lesions identified by each script compared against the lesions in Microbigg-E? 

- There were 624 records were a mutation was found for the same gene
  - 403 (64.6%) of those records show the same lesion from `DGW.py` and `scanner.R`
  - 61 (9.8%) of those records show the same lesion in `Microbigg-E` and either of the DGW scripts

```{r}
DT::datatable(dgw_combined_outputs %>% 
  filter((mb_element_symbol == p_element_symbol | mb_element_symbol == r_element_symbol) & (mb_lesion != p_lesion | mb_lesion != r_lesion)) %>% 
  select(asm_acc, 
         element_symbol = mb_element_symbol,
         mb_lesion, 
         p_lesion,
         r_lesion)
  )

# dgw_combined_outputs %>% 
#   filter((mb_element_symbol == p_element_symbol | mb_element_symbol == r_element_symbol) & (mb_lesion != p_lesion | mb_lesion != r_lesion)) %>% 
#   summarize(same_dgw_lesion = n_distinct(case_when(p_lesion == r_lesion ~ record_id)),
#             same_mb_lesion = n_distinct(case_when(mb_lesion == r_lesion | mb_lesion == p_lesion ~ record_id))
#   )

```

### How many lesions in Microbigg-E are not detected by the DGW scripts? (show lesions)

- Out of the 6406 mutations present in Microbigg-E, 4769 (74.4%) were missed by both DGW scripts
  - `DGW.py` missed 4800 (74.9%)  Microbigg-E mutations
  - `scanner.R` missed 5037 (78.6%) Microbigg-E mutations

```{r}
DT::datatable(dgw_combined_outputs %>% 
  summarize(in_microbigge = n_distinct(case_when(!is.na(mb_element_symbol) ~ record_id), na.rm = TRUE),
            missed_from_microbigge = n_distinct(case_when(!is.na(mb_element_symbol) & (is.na(p_element_symbol) & is.na(r_element_symbol)) ~ record_id), na.rm = TRUE),
            p_missed_from_microbigge = n_distinct(case_when(!is.na(mb_element_symbol) & is.na(p_element_symbol) ~ record_id), na.rm = TRUE),
            r_missed_from_microbigge = n_distinct(case_when(!is.na(mb_element_symbol) & is.na(r_element_symbol) ~ record_id), na.rm = TRUE)
  )
)

```
```{r, echo = FALSE}
union(dgw_combined_outputs %>% 
  filter(!is.na(mb_element_symbol)) %>% 
  mutate(result = case_when(is.na(p_mutation_id) ~ "Missed by DGW",
                            !is.na(p_mutation_id) ~ "Found mutations"
                            )) %>% 
  group_by(result) %>% 
  summarize(script = "DWG.py",
            mutations = n_distinct(record_id))
  ,
dgw_combined_outputs %>% 
  filter(!is.na(mb_element_symbol)) %>% 
  mutate(result = case_when(is.na(r_mutation_id) ~ "Missed by DGW",
                            !is.na(r_mutation_id) ~ "Found mutations"
                            )) %>% 
  group_by(result) %>% 
  summarize(script = "scanner.R",
            mutations = n_distinct(record_id))
) %>% 
  ggplot(., aes(fill=result, y=mutations, x=script)) + 
    geom_bar(position="stack", stat="identity", width=0.7) +
    scale_fill_manual(name="",values=c('#68a1b3', '#a65005')) +
    theme(
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.border = element_blank(),
      panel.background = element_blank(),
      text = element_text(size = 16)) +
    ylab("# of Mutations") + xlab("Script") +
    scale_x_discrete(labels = function(x) stringr::str_wrap(x, width = 20))
  
```

```{r}
DT::datatable(dgw_combined_outputs %>% 
               filter (!is.na(mb_element_symbol) & (is.na(p_element_symbol) | is.na(r_element_symbol))) %>% 
               select(asm_acc, mb_element_symbol, mb_lesion, p_lesion, r_lesion)
  )

```


### How many of the detections in Microbigg-E missed by DWG should have been detected? (i.e. how many are nonsense and frameshift mutations)

As indicated above, there were 4769 records were a mutation was reported in Microbigg-E, but no lesion was detected by the DGW scripts. Looking at the specific lesions, we find that those records account for 23 unique lesions, of which 6 are nonsense mutations (early STOP codon), reported in 15 records in the testset. 

```{r}
DT::datatable(dgw_combined_outputs %>%
  filter (!is.na(mb_element_symbol) & ((is.na(p_element_symbol) | is.na(r_element_symbol)))) %>% 
  group_by(mb_element_symbol, mb_lesion) %>% 
  summarize(records = n_distinct(record_id, na.rm = TRUE)) %>% 
    arrange(desc(records))
)

```

However, this does not mean that the DGW scripts never found those mutations. So next we look into the specific lesions for each gene and find that only 18 of the lesions present in Microbigg-E are never found by DGW, of which none are nonsense mutations. 

The table below shows all the lesions that ar reported in Microbigg-E but were not found by either of the DGW scripts. The `records` field shows the records where the lesion is reported in Mocrobigg-E for the entire testset, and the `other_lesions` are records where DGW found a different lesion than the one reported in Microbigg-E.

```{r}
# All lesions reported in Microbigg-E for isolates tested
## Value of 1 in missing field means that there are records where DGW scripts missed the mutation
mb_lesions <- dgw_combined_outputs %>%
  filter(!is.na(mb_element_symbol)) %>% 
  mutate(missing= case_when((is.na(p_element_symbol) | is.na(r_element_symbol)) | 
                              paste0(tolower(mb_element_symbol), mb_lesion) != paste0(tolower(p_element_symbol), p_lesion) |
                              paste0(tolower(mb_element_symbol), mb_lesion) != paste0(tolower(r_element_symbol), r_lesion)
                              ~ 1, .default = 0),
         mb_gene_lesion = paste0(tolower(mb_element_symbol), "_", mb_lesion)
         ) %>% 
  group_by(mb_gene_lesion, missing) %>% 
  summarize(records = n_distinct(record_id)) %>% 
  arrange(mb_gene_lesion, missing)

# Missing lesions reporting a different lesion in DGW from what is reported in Microbigg-E
other_lesion <- dgw_combined_outputs %>% 
  filter(!is.na(mb_element_symbol)) %>% 
  mutate(mb_gene_lesion = paste0(tolower(mb_element_symbol), "_", mb_lesion)) %>% 
  group_by(mb_gene_lesion) %>% 
  summarize(other_lesions = n_distinct(case_when(!is.na(p_element_symbol) | !is.na(r_element_symbol) ~ record_id)))
  

# All lesions reported in Microbigg-E that DGW models didn't find for any record
DT::datatable(mb_lesions %>% 
  filter(max(row_number()) == 1) %>% 
  filter(missing == 1) %>% 
  left_join(other_lesion, join_by(mb_gene_lesion == mb_gene_lesion)) %>% 
  select(mb_gene_lesion, records, other_lesions) %>% 
  arrange(desc(records)))

```

All the lesions indicated above were confirmed to be insertions or point mutations that did not lead to a frameshift or early stop codon, with the only exception being `cira_Y253CfsTer5` which was identified by the DGW scripts as `cira_Y253Cfs`. Therefore, the DGW scripts did not miss any of the nonsense and frameshift mutations identified in `Microbigg-E`. However, there were 11 lesions that the DGW scripts did not identify consistently.

The following table shows lesions inconsistently identified by DGW scripts, which are detected for some records by the DGW scripts, but missed in some records where the lesion is reported in Microbigg-E. All of these lesions correspond to nonsense mutations, so the DGW scripts are successful in finding them, but not consistent across the testset. 


```{r}
# Mutations reported in Microbigg-E that DGW missed for some records 
# Records field indicates the number of records where DGW scripts missed the mutation
# These are the mutations that DGW should have found, so we need to understand why they were missed

# DT::datatable(mb_lesions %>% 
#   filter(max(row_number(mb_gene_lesion)) == 2) %>% 
#   mutate(dgw_found = max(case_when(missing == 0 ~ records), na.rm = TRUE),
#          dgw_missed = max(case_when(missing == 1 ~ records), na.rm = TRUE)
#          ) %>% 
#   select(mb_gene_lesion, dgw_found, dgw_missed) %>% 
#   unique())

```

```{r}
DT::datatable(
dgw_combined_outputs %>%
  filter(!is.na(mb_element_symbol)) %>% 
  mutate(p_missing= case_when(is.na(p_element_symbol) | 
                              paste0(tolower(mb_element_symbol), mb_lesion) != paste0(tolower(p_element_symbol), p_lesion)
                              ~ 1, .default = 0),
         r_missing= case_when(is.na(r_element_symbol) | 
                              paste0(tolower(mb_element_symbol), mb_lesion) != paste0(tolower(r_element_symbol), r_lesion)
                              ~ 1, .default = 0),
         mb_gene_lesion = paste0(tolower(mb_element_symbol), "_", mb_lesion)
         ) %>%
  group_by(mb_gene_lesion) %>% 
  summarize(dgw_found = n_distinct(case_when(p_missing == 0 | r_missing == 0 ~ record_id), na.rm = TRUE),
            p_missed_records = n_distinct(case_when(p_missing == 1 ~ record_id), na.rm = TRUE),
            r_missed_records = n_distinct(case_when(r_missing == 1 ~ record_id), na.rm = TRUE)
            ) %>% 
  unique() %>% 
  filter(dgw_found > 0 & (p_missed_records > 0 | r_missed_records > 0))
)

```


Looking into each of these lesions:

- ompk36_Q313STOP:
  - `DGW.py` found STOP mutations in positions 314-320 for the reported 14 missed records
- oprd_W277STOP:
  - `DGW.py` found the mutation in all records. The reported 2 missing records found `W277STOP,S401fs`
- oprd_W417STOP:
  - `DGW.py` found the mutation in all records. The reported missing record found `W422STOP,E171fs`
- oprd_Q142STOP:
  - True `DGW.py` miss. No mutation detected. 
- oprd_S319STOP:
  - `DGW.py` found `S321STOP,S59fs,D61fs` in the reported missed record.





