---
title: "Dead Gene Walking Results"
output: html_document
date: "2024-09-26"
---

```{r setup, include = FALSE}
# Set working directory
knitr::opts_knit$set(root.dir = '~/Documents/Projects/resistance_is_futile/amr-2024-team-prasad-feldgarden')
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
```

# Introduction

This notebook explores the overlap across the results obtained in the codaton by our tool dead-gene-walking (DGW) against MicroBIGG-E.

Comparing DGW results against MicroBIGG-E:

- How many of the mutations DGW identifies are reported in MicroBIGG-E?
- How many genome accessions are present in MicroBIGG-E but do not report the mutations DGW found?

In addition:

- How prevalent are these mutations (i.e. group by lession_type and ref_gene in our output)


# Setup

A random set of 3000 genome accession IDs for *Acinetobacter baumannii*, *Pseudomonas aeruginosa* and *Klebsiella pneumoniae* (1000 isolates for each) were obtained using NCBI's `datasets` as described [here](https://github.com/NCBI-Codeathons/amr-2024-team-prasad-feldgarden/tree/main/data/randomset1#test-on-a-random-set-of-assemblies).


## MicroBIGG-E subset

A subset of the data available in MicroBIGG-E for the selected isolates was obtained from BigQuery and stored in the repository [here](https://github.com/NCBI-Codeathons/amr-2024-team-prasad-feldgarden/blob/main/results/microbigge_subset_randomset.csv). 

## DGW results

DGW was ran as described [here](https://github.com/NCBI-Codeathons/amr-2024-team-prasad-feldgarden/tree/main/script), and the output for each taxon is shown in individual tables in the repository [here](https://github.com/NCBI-Codeathons/amr-2024-team-prasad-feldgarden/tree/main/results). These results are going to be compared against the MicroBIGG-E subset in the following section. 

# Analysis of results

```{r}
# Install packges
library(DBI)
library(duckdb)
library(tidyverse)
library(dplyr)


# Import data
Kp_genomes <- read_tsv("results/Kp_genomes.dgw")
Pa_genomes <- read_tsv("results/Pa_genomes.dgw")
Ab_genomes <- read_tsv("results/Ab_genomes.dgw")

microbigge <-  read_csv("results/microbigge_subset_randomset.csv")

# Set up in-memory database for modeling
con <- dbConnect(duckdb(), dbdir = ":memory:")

# Register dfs into database
dbWriteTable(con, "kp_genomes", Kp_genomes)
dbWriteTable(con, "pa_genomes", Pa_genomes)
dbWriteTable(con, "ab_genomes", Ab_genomes)

dbWriteTable(con, "microbigge", microbigge)

```

Let's first prepare the data to establish the comparisons/analysis.

```{r}
# Combine DGW results
dgw_combined <- dbGetQuery(con,
"SELECT 
  *,
  md5(CONCAT(element_symbol, asm_acc, contig_acc, lesion)) AS mutation_id,
  'kp_genomes' AS cohort
FROM kp_genomes
UNION ALL 
SELECT 
  *,
  md5(CONCAT(element_symbol, asm_acc, contig_acc, lesion)) AS mutation_id,
  'pa_genomes' AS cohort
FROM pa_genomes 
UNION ALL 
SELECT 
  *,
  md5(CONCAT(element_symbol, asm_acc, contig_acc, lesion)) AS mutation_id,
  'ab_genomes' AS cohort
FROM ab_genomes
")
dbWriteTable(con, "dgw_combined", dgw_combined, overwrite = TRUE)


# Is DGW reported in MicroBIGG-E?
dgw_in_mbigge <- dbGetQuery(con,
"SELECT
  dgw_combined.element_symbol,
  dgw_combined.asm_acc,
  dgw_combined.contig_acc,
  dgw_combined.lesion,
  dgw_combined.lesion_type,
  dgw_combined.cohort,
  dgw_combined.mutation_id,
  microbigge.element_symbol AS mb_element_symbol,
  microbigge.class
FROM dgw_combined
LEFT JOIN microbigge
ON dgw_combined.asm_acc = microbigge.asm_acc
AND LOWER(microbigge.element_symbol) LIKE CONCAT('%', LOWER(dgw_combined.element_symbol),'%')
")


#Find new mutations against MicroBIGG-E mutations
dgw_new <- dbGetQuery(con,
"WITH match AS (
SELECT DISTINCT
  mutation_id
FROM dgw_combined
WHERE LOWER(microbigge.element_symbol) LIKE CONCAT('%', LOWER(dgw_combined.element_symbol),'%')
)
SELECT DISTINCT
  dgw_combined.element_symbol,
  dgw_combined.asm_acc,
  dgw_combined.contig_acc,
  dgw_combined.lesion,
  dgw_combined.lesion_type,
  dgw_combined.cohort,
  STRING_AGG(microbigge.element_symbol) OVER (PARTITION BY dgw_combined.element_symbol, dgw_combined.asm_acc) AS mb_element_symbol,
FROM dgw_combined
LEFT JOIN microbigge
ON dgw_combined.asm_acc = microbigge.asm_acc
LEFT JOIN match
USING(mutation_id)
WHERE  match.mutation_id IS NULL
")

```


## How many of the mutations DGW identifies are reported in MicroBIGG-E?

Out of the 3000 isolates analyzed by DGW, 877 unique mutations were found across 789 isolates. 
  - 150 mutations were also present in MicroBIGG-E

```{r}
#Total output
dgw_output_combined <- dgw_combined %>% 
  summarize(mutations = n_distinct(mutation_id, na.rm = TRUE),
            isolates = n_distinct(asm_acc, na.rm = TRUE),
            contigs = n_distinct(contig_acc, na.rm = TRUE))


# question1
q1 <- dgw_in_mbigge %>% 
  filter(!is.na(mb_element_symbol)) %>% 
  group_by(element_symbol) %>% 
  summarize(mutations = n_distinct(mutation_id),
            isolates = n_distinct(asm_acc)) %>% 
  arrange(desc(mutations))

# q1_inverse <- dgw_in_mbigge %>% 
#   filter(is.na(mb_element_symbol)) %>% 
#   group_by(element_symbol) %>% 
#   summarize(mutations = n_distinct(mutation_id),
#             isolates = n_distinct(asm_acc))


print(q1)

```



## How many isolates are present in MicroBIGG-E but do not report the mutations DGW found?

DGW found 727 new mutations in 353 isolates. 

*Klebsiella pneumoniae* OmpK35_N29fs was the must abundant, found in 131 isolates.

```{r}
# q2 
q2 <- dgw_in_mbigge %>% 
  filter(is.na(mb_element_symbol)) %>% 
  mutate(phenotype = case_when(grepl("resistant", cohort) ~ "Resistant",
                               grepl("sensitive", cohort) ~ "Sensitive"),
         taxon = case_when(grepl("kp", cohort) ~ "Klebsiella pneumoniae",
                           grepl("pa", cohort) ~ "Pseudomonas aeruginosa",
                           grepl("ab", cohort) ~ "Acinetobacter baumannii")) %>% 
  group_by(phenotype, taxon, element_symbol, lesion, lesion_type) %>% 
  summarize(mutations = n_distinct(mutation_id),
            isolates = n_distinct(asm_acc)) %>% 
  arrange(desc(mutations))

print(q2)

```


```{r}
# proportion of new mutations
q2 %>% 
  group_by(lesion_type, taxon) %>% 
  summarize(mutations = sum(mutations)) %>% 
  ggplot(., aes(fill=lesion_type, y=mutations, x=taxon)) + 
    geom_bar(position="stack", stat="identity") +
    scale_fill_manual(name="",values=c('#347d38', '#c4a208', '#2596be')) +
    theme(
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.border = element_blank(),
      panel.background = element_blank()) +
    ylab("# of Mutations") + xlab("")

```


```{r}
# q3
q3 <- dgw_combined %>% 
  mutate(phenotype = case_when(grepl("resistant", cohort) ~ "Resistant",
                               grepl("sensitive", cohort) ~ "Sensitive"),
         taxon = case_when(grepl("kp", cohort) ~ "Klebsiella pneumoniae",
                           grepl("pa", cohort) ~ "Pseudomonas aeruginosa",
                           grepl("ab", cohort) ~ "Acinetobacter baumannii")) %>% 
  group_by(phenotype, taxon, lesion) %>% 
  summarize(mutations = n_distinct(mutation_id),
            isolates = n_distinct(asm_acc))


# Split of DGW mutations in MicroBIGG-E vs New
dgw_in_mbigge %>% 
  mutate(split = case_when(is.na(mb_element_symbol) ~ "New",
                          !is.na(mb_element_symbol) ~ "In MicroBiGG-E"),
         taxon = case_when(grepl("kp", cohort) ~ "Klebsiella pneumoniae",
                           grepl("pa", cohort) ~ "Pseudomonas aeruginosa",
                           grepl("ab", cohort) ~ "Acinetobacter baumannii")) %>% 
  group_by(split, taxon) %>% 
  summarize(mutations = n_distinct(mutation_id),
            isolates = n_distinct(asm_acc)) %>% 
  ggplot(., aes(fill=split, y=mutations, x=taxon)) + 
    geom_bar(position="stack", stat="identity") +
    scale_fill_manual(name="",values=c('#a65005', '#68a1b3')) +
    theme(
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.border = element_blank(),
      panel.background = element_blank()) +
    ylab("# of Mutations") + xlab("")

```


```{r}
# q3.2
q3_2 <- dgw_combined %>% 
  group_by(lesion_type) %>% 
  summarize(mutations = n_distinct(mutation_id),
            isolates = n_distinct(asm_acc))

dgw_combined %>% 
  mutate(taxon = case_when(grepl("kp", cohort) ~ "Klebsiella pneumoniae",
                           grepl("pa", cohort) ~ "Pseudomonas aeruginosa",
                           grepl("ab", cohort) ~ "Acinetobacter baumannii")) %>% 
  group_by(lesion_type, taxon) %>% 
  summarize(mutations = n_distinct(mutation_id),
            isolates = n_distinct(asm_acc)) %>% 
  ggplot(., aes(fill=lesion_type, y=mutations, x=taxon)) + 
    geom_bar(position="stack", stat="identity", width=0.7) +
    scale_fill_manual(name="",values=c('#347d38', '#c4a208', '#2596be')) +
    theme(
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.border = element_blank(),
      panel.background = element_blank()) +
    ylab("# of Mutations") + xlab("") 

  
```

# Summary
    
- DGW identified 877 mutations out of the 3000 isolates analyzed
  - 150 mutations were present in MicroBIGG-E too
  - 727 new mutations in 353 isolates
      
  